### PR TITLE
Allow multiple internal mount parameters

### LINKED ISSUE #479: Log if mount is being manually run by user or kicked off by the service
We are seeing several errors come thru that say 'disk I/O error'

The two flavors are:
Failed to create src folder callback listener Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 10: 'disk I/O error'.
System.IO.IOException: The device is not ready.
The first is almost always on startup.  The second is sometimes on startup.  

Reaching out to a few users they just run mount or open/close razzle.  We'd like to get a better idea of this failure.  One hypothesis is that the service is kicking off mount when it can't run, and adding this data will help field that theory.  In general it's also helpful to know if users are having to manually mount versus the service.

Since the service kicks off "mount", this makes it tricky to determine how the start occurred.
Options discussed:
1. Add an Env variable in the service, that MountVerb will look for and log to determine start
2. Add an new optional parameter on Mount
@sanoursa, let me know your preference here on approach

### PR DESCRIPTION
**Pull request by jeschu1**
We are seeing several errors come thru that say 'disk I/O error'

**The two flavors are:**
1. Failed to create src folder callback listener Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 10: 'disk I/O error'.
2. System.IO.IOException: The device is not ready.
The first is almost always on startup. The second is sometimes on startup.

This change logs whether the mount is started by the service.  This will help us determine how often mount is failing when launched via the service.

**There were 3 evaluated approaches**
1. Walk the process tree
This was decided against since it's not cross-plat and walking process trees is non-trivial
2. Use an env var
This was decided against since we start processes using native code and adding an env variable there adds new complexity.  It's also less than ideal since we don't pass any other arguments this way.
3. Use the existing internal-only property on MountVerb
I extended this option to allow you send a series of var=val;var=val; type arguments.  I originally considered using json but passing args via json has it's own set of complexities (double escaping) and can become quite verbose.  Since this is internal I'm voting to keep it simple.

The third approach here is what I went with.  It seems the most standard as well as giving an easy way to add internal parameters in the future.

resolves #479

### CODE REVIEW

**Comment by jrbriggs** [comment]
@jeschu1 can you summarize the motivation for this change in the description?

**Comment by jeschu1** [comment]
I put the PR up to double check the tests.  Once they pass I'll update the context.  Please don't dive into this PR yet friends!

File: GVFS/GVFS.Common/GVFSConstants.cs [hunk 1]
```diff
@@ -210,7 +210,9 @@ public static class VerbParameters
         {
             public static class Mount
             {
-                public const string ServiceName = "internal_use_only_service_name";
+                public const string InternalParameters = "internal_use_only_parameters";
```
Comment (jeschu1) [inline GVFS/GVFS.Common/GVFSConstants.cs]: I'm not sure if changing internal_use_only_service_name to internal_use_only_parameters is a legit thing to do.  If anyone was using the parameter (which they shouldn't it's internal :-) ) they'd be broken.  I do like the new name as a better description though.

File: GVFS/GVFS/CommandLine/GVFSVerb.cs [hunk 2]
```diff
@@ -34,12 +34,41 @@ public GVFSVerb(bool validateOrigin = true)
...
 
         [Option(
-            GVFSConstants.VerbParameters.Mount.ServiceName,
-            Default = GVFSConstants.Service.ServiceName,
+            GVFSConstants.VerbParameters.Mount.InternalParameters,
             Required = false,
-            HelpText = "This parameter is reserved for internal use.")]
+            HelpText = "This parameter is reserved for internal use." +
+                " Accepts input in the form key1=value;key2=value;key3=value etc..")]
+        public string InternalParameters
```
Comment (jeschu1) [inline GVFS/GVFS/CommandLine/GVFSVerb.cs]: If we give this approach the thumbs up, I'll add unit tests for the below

File: GVFS/GVFS.Mount/InProcessMountVerb.cs [hunk 3]
```diff
@@ -47,6 +47,14 @@ public InProcessMountVerb()
...
         public bool ShowDebugWindow { get; set; }
 
+        [Option(
+            's',
+            GVFSConstants.VerbParameters.Mount.StartedByService,
+            Default = "false",
+            Required = false,
+            HelpText = "Service initiated mount.")]
+        public string StartedByService { get; set; }
+
```
Comment (jeschu1) [inline GVFS/GVFS.Mount/InProcessMountVerb.cs]: Arguably this isn't consistent with MountVerb which allows you to pass a series of internal parameters.  On the InProcessMountVerb itself, you only need 1 parameter so I'd rather not make it take a series until/unless the need arises.

Comment (sanoursa) [reply GVFS/GVFS.Mount/InProcessMountVerb.cs]: The `MountVerb` gets called in-proc by other verbs. No such requirement here.

File: GVFS/GVFS.Common/GVFSConstants.cs (hunk 1 above)
Comment (sanoursa) [inline GVFS/GVFS.Common/GVFSConstants.cs]: Not your doing, but this parameter doesn't belong in `VerbParameters.Mount`, since it's not specific to just `MountVerb` (I believe it was when it was first added, and later got moved to all verbs).

I would create a class named `VerbParameters.Internal` with a member named `Name` and other members for each of the parameters that it can contain.

Comment (sanoursa) [inline GVFS/GVFS.Common/GVFSConstants.cs]: Minor, but I would name it just "internal_use_only" and drop the "_parameters". Or even just "internal_use". Make the CLI, especially when captured in Task Manager, etc, slightly less verbose.

Comment (sanoursa) [reply GVFS/GVFS.Common/GVFSConstants.cs]: Or even better, define a class with properties, and use Json to serialize/deserialize its members. There's no need for string-typed values here since we don't need any cross-version compatibility.

File: GVFS/GVFS.Mount/InProcessMountVerb.cs [hunk 4]
```diff
@@ -47,6 +47,14 @@ public InProcessMountVerb()
             HelpText = "Show the debug window.  By default, all output is written to a log file and no debug window is shown.")]
         public bool ShowDebugWindow { get; set; }
 
+        [Option(
+            's',
```
Comment (sanoursa) [inline GVFS/GVFS.Mount/InProcessMountVerb.cs]: No need for a short name, since no user will be running this command

Comment (sanoursa) [reply GVFS/GVFS.Mount/InProcessMountVerb.cs]: (I know that is inconsistent with the other params here, but pretty much all of them are holdovers from the early days and should be removed anyways)

File: GVFS/GVFS/CommandLine/GVFSVerb.cs [hunk 5]
```diff
@@ -34,12 +34,41 @@ public GVFSVerb(bool validateOrigin = true)
         public abstract string EnlistmentRootPathParameter { get; set; }
 
         [Option(
-            GVFSConstants.VerbParameters.Mount.ServiceName,
-            Default = GVFSConstants.Service.ServiceName,
+            GVFSConstants.VerbParameters.Mount.InternalParameters,
             Required = false,
-            HelpText = "This parameter is reserved for internal use.")]
+            HelpText = "This parameter is reserved for internal use." +
+                " Accepts input in the form key1=value;key2=value;key3=value etc..")]
```
Comment (sanoursa) [inline GVFS/GVFS/CommandLine/GVFSVerb.cs]: I would drop this. Users don't need to know or care what the format is.

File: GVFS/GVFS/CommandLine/GVFSVerb.cs [hunk 6]
```diff
@@ -34,12 +34,41 @@ public GVFSVerb(bool validateOrigin = true)
...
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
         public string ServiceName { get; set; }
 
+        public string StartedByService { get; set; }
```
Comment (sanoursa) [inline GVFS/GVFS/CommandLine/GVFSVerb.cs]: This should be a bool, not a string. But this is moot if you parse the json into a strongly-typed `InternalParameters` instance.

File: GVFS/GVFS/CommandLine/GVFSVerb.cs [hunk 7]
```diff
@@ -34,12 +34,41 @@ public GVFSVerb(bool validateOrigin = true)
...
+            {
+                if (!string.IsNullOrEmpty(value))
+                {
+                    string[] values = value.Split(';');
+                    foreach (string s in values)
+                    {
+                        string[] pair = s.Split('=');
+                        if (pair.Length == 2)
+                        {
+                            if (string.Equals(pair[0], GVFSConstants.VerbParameters.Mount.StartedByService))
```
Comment (mjcheetham) [inline GVFS/GVFS/CommandLine/GVFSVerb.cs]: Do we care about case-sensitivity? I'd use `StringComparer.OrdinalIgnoreCase.Equals` here.

File: GVFS/GVFS.Service/GVFSMountProcess.cs [hunk 8]
```diff
@@ -58,7 +58,10 @@ public void Dispose()
 
         private bool CallGVFSMount(string repoRoot)
         {
-            return this.CurrentUser.RunAs(Configuration.Instance.GVFSLocation, "mount " + repoRoot);
+            return this.CurrentUser.RunAs(
+                Configuration.Instance.GVFSLocation,
```
Comment (mjcheetham) [inline GVFS/GVFS.Service/GVFSMountProcess.cs]: This expression is now quite long with lots of string appending. Makes it hard to see if the spaces are in the correct place to separate arguments/values.

Consider making this an `$"interpolated {string}"` (or use a `StringBuilder`) and pull the final formatted string out of the `RunAs` call.

File: GVFS/GVFS.FunctionalTests/Tools/GVFSProcess.cs [hunk 9]
```diff
@@ -94,7 +94,7 @@ private string CallGVFS(string args, bool failOnError = false, string trace = nu
         {
             ProcessStartInfo processInfo = null;
             processInfo = new ProcessStartInfo(this.pathToGVFS);
-            processInfo.Arguments = args + " --internal_use_only_service_name " + GVFSServiceProcess.TestServiceName;
+            processInfo.Arguments = args + " --internal_use_only_parameters " + GVFSHelpers.GetTestServiceParamater();
```
Comment (mjcheetham) [inline GVFS/GVFS.FunctionalTests/Tools/GVFSProcess.cs]: Should we (or have we already) pull out the name of the parameter into a constant somewhere?
```csharp
const string InternalUseParametersName = "internal_use_only_parameters";
```

File: GVFS/GVFS.Common/GVFSConstants.cs (hunk 1 above)
Comment (jeschu1) [reply GVFS/GVFS.Common/GVFSConstants.cs]: I updated it to be a serialized Json class.  It's a bit more painful (supplying json across a command line is no fun!), but that will probably encourage folks not to use it which is what we want, while forcing type.

File: GVFS/GVFS/CommandLine/GVFSVerb.cs (hunk 7 above)
Comment (jeschu1) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: I updated it to be a serialized class.  No ambiguity now and type safety.

File: GVFS/GVFS.Service/GVFSMountProcess.cs [hunk 10]
```diff
@@ -58,7 +58,10 @@ public void Dispose()
 
         private bool CallGVFSMount(string repoRoot)
         {
-            return this.CurrentUser.RunAs(Configuration.Instance.GVFSLocation, "mount " + repoRoot);
+            MountInternal mountInternal = new MountInternal() { StartedByService = true };
+            return this.CurrentUser.RunAs(
+                Configuration.Instance.GVFSLocation,
+                $"mount {repoRoot} --internal_use_only {mountInternal.ToJson()}");
```
Comment (wilbaker) [inline GVFS/GVFS.Service/GVFSMountProcess.cs]: Use `GVFSConstants.VerbParameters.Mount.InternalUseOnly` here rather than `internal_use_only`

File: GVFS/GVFS/CommandLine/GVFSVerb.cs [hunk 11]
```diff
@@ -34,12 +34,30 @@ public GVFSVerb(bool validateOrigin = true)
...
+                    }
+
+                    this.StartedByService = mountInternal.StartedByService;
+                }
+            }
+        }
+
         public string ServiceName { get; set; }
 
+        public bool StartedByService { get; set; }
```
Comment (wilbaker) [inline GVFS/GVFS/CommandLine/GVFSVerb.cs]: Minor: Can `ServiceName` and `StartedByService` be changed to `protected`?

(It actually looks like `Unattended`, `ServicePipeName`, and `Output` can probably be `protected` as well)

File: GVFS/GVFS.Common/MountInternal.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
+﻿using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace GVFS.Common
+{
+    public class MountInternal
```
Comment (sanoursa) [inline GVFS/GVFS.Common/MountInternal.cs]: This currently sounds like a method name rather than a class name, and also these parameters apply to all verbs, not just mount. Rename to `InternalVerbParameters`?

File: GVFS/GVFS.Common/MountInternal.cs [hunk 13]
```diff
@@ -0,0 +1,23 @@
+﻿using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Text;
```
Comment (sanoursa) [inline GVFS/GVFS.Common/MountInternal.cs]: Remove unnecessary using statements

File: GVFS/GVFS/CommandLine/GVFSVerb.cs [hunk 14]
```diff
@@ -34,12 +34,30 @@ public GVFSVerb(bool validateOrigin = true)
...
+                    {
+                        this.ServiceName = mountInternal.ServiceName;
+                    }
+
+                    this.StartedByService = mountInternal.StartedByService;
+                }
+            }
+        }
+
         public string ServiceName { get; set; }
```
Comment (sanoursa) [inline GVFS/GVFS/CommandLine/GVFSVerb.cs]: Make these setters private

File: GVFS/GVFS.Common/GVFSConstants.cs [hunk 15]
```diff
@@ -210,7 +210,8 @@ public static class VerbParameters
         {
             public static class Mount
             {
-                public const string ServiceName = "internal_use_only_service_name";
+                public const string InternalUseOnly = "internal_use_only";
+                public const string StartedByService = "StartedByService";
```
Comment (sanoursa) [inline GVFS/GVFS.Common/GVFSConstants.cs]: I don't think this is being used any more?

File: GVFS/GVFS.Common/GVFSConstants.cs [hunk 16]
```diff
@@ -210,7 +210,8 @@ public static class VerbParameters
         {
             public static class Mount
             {
-                public const string ServiceName = "internal_use_only_service_name";
+                public const string InternalUseOnly = "internal_use_only";
```
Comment (sanoursa) [inline GVFS/GVFS.Common/GVFSConstants.cs]: Move this up to the parent class, since it's not specific to Mount anymore

File: GVFS/GVFS/CommandLine/GVFSVerb.cs (hunk 11 above)
Comment (sanoursa) [reply GVFS/GVFS/CommandLine/GVFSVerb.cs]: They can be, but then so could everything else in these classes. This is the top of the stack so there isn't any other code calling these methods/properties other than the verbs themselves.

File: GVFS/GVFS.Common/MountInternal.cs (hunk 12 above)
Comment (sanoursa) [reply GVFS/GVFS.Common/MountInternal.cs]: Also does it belong here in `GVFS.Common`? I think it could probably move into the CLI project instead.

File: GVFS/GVFS.Common/MountInternal.cs [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+namespace GVFS.Common
+{
+    public class MountInternal
+    {
+        public string ServiceName { get; set; }
+        public bool StartedByService { get; set; }
+
+        public static MountInternal FromJson(string json)
+        {
+            return JsonConvert.DeserializeObject<MountInternal>(json);
```
Comment (sanoursa) [inline GVFS/GVFS.Common/MountInternal.cs]: Can this throw if the json is not in the right format?

Comment (jeschu1) [reply GVFS/GVFS.Common/MountInternal.cs]: @sanoursa yes it can throw.  Since it's internal I'm thinking I'd prefer the mount not to start if we get bad input.

Comment (sanoursa) [reply GVFS/GVFS.Common/MountInternal.cs]: Yes, agreed that it should take down the process. Please just confirm that it is getting handled and logged as it tears down the process.

File: GVFS/GVFS.Common/MountInternal.cs (hunk 12 above)
Comment (jeschu1) [reply GVFS/GVFS.Common/MountInternal.cs]: @sanoursa CLI project?  Sorry i'm not seeing it.  It's used by GVFS.Service/GVFS.Windows(Mac) which is why I was thinking Common might be a good home.

File: GVFS/GVFS.Common/MountInternal.cs (hunk 13 above)
Comment (jeschu1) [reply GVFS/GVFS.Common/MountInternal.cs]: Done, but it's a little spooky you could spot that from just a PR :-)

File: GVFS/GVFS.Common/MountInternal.cs (hunk 12 above)
Comment (sanoursa) [reply GVFS/GVFS.Common/MountInternal.cs]: `GVFS.Windows` and `GVFS.Mac` are what I meant by the CLI project(s), since they produce the CLI executable.

This class exists for interacting with the CLI, which is why I suggested putting it there. `GVFS.Service` should be able to take a dependency on `GVFS.Windows`.

Comment (derrickstolee) [reply GVFS/GVFS.Common/MountInternal.cs]: You're right, this is a class specifically for `MountVerb`.

It should also be used by whatever Mac equivalent of the `GVFS.Service` project will be. I imagine that the Mac service can take a dependence on `GVFS.Mac` and get the same functionality. Is that right?

Comment (sanoursa) [reply GVFS/GVFS.Common/MountInternal.cs]: It's not specifically for `MountVerb`, it gets used by all CLI verbs.

And yes the Mac equivalent of the service should be able to depend on the CLI project as well.

File: GVFS/GVFS.Common/GVFSConstants.cs (hunk 15 above)
Comment (jeschu1) [reply GVFS/GVFS.Common/GVFSConstants.cs]: @sanoursa  Where would you recommend adding the StartedByService Constant?
It's used in both InProcessMountVerb (GVFS.Win) and MountVerb (GVFS.Mount.Win)

File: GVFS/GVFS.Common/MountInternal.cs (hunk 12 above)
Comment (jeschu1) [reply GVFS/GVFS.Common/MountInternal.cs]: @sanoursa I tried to add a dependency on GVFS.Windows from GVFS.Service.

However since GVFS.Service has a dependency on GVFS.Service.UI which has a shared file ( PlatformLoader.Windows.cs), I'm getting build errors saying it doesn't know which GVFSPlatformLoader to use.  Not sure how to resolve this one.

Comment (jeschu1) [reply GVFS/GVFS.Common/MountInternal.cs]: I tried removing the shared file, to no avail.  As you've previously mentioned GVFS.Service.UI needs some attention in terms of refactoring / getting it to work again.  I'm going to leave this final change out of scope for the PR.  But hopefully we prioritize the notifications soon and then we can clean this up.
